src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,153 @@
+    const parentNode = _.get(tree, parentPath)
+    const grandParentNode = _.get(tree, grandParentPath)
+    const nodeToMergeIntoPath = (parentNode && Object.keys(parentNode).length !== 0) ? parentPath : ((grandParentNode && Object.keys(grandParentNode).length - 1 > 0) ? grandParentPath : false)
+    if (nodeToMergeIntoPath) {
+      // checking all the nodes from the direct parent or grandparent , if available updating all the descendants lastUpdated
+      if (commonPath.length - nodeToMergeIntoPath.length === 2) _.unset(tree, parentPath) // if the parent path is merged to grandParent then deleting the parent node
+      findTreeDescendants(tree, nodeToMergeIntoPath).forEach(descendant => {
+        _.set(tree, [EM_TOKEN, ...pathToContext(descendant.path)], { leaf: true, lastUpdated: timestampUpdate ? timestamp() : descendant.lastUpdated, path: descendant.path })
+      })
+    }
+    // if cannot be merged to either parent or grandParent then just making parent path as the leaf node *** exception if parent path is __ROOT__ node ***
+    else if (parentPath.length > 1) _.set(tree, parentPath, { leaf: true, lastUpdated: timestamp(), path: oldPath.slice(0, parentPath.length - 1) })
+  }
+}
+
+const nodeMove = (tree, oldPath, newPath) => {
+  const oldContext = [EM_TOKEN, ...pathToContext(oldPath)]
+  const newContext = [EM_TOKEN, ...pathToContext(newPath)]
+
+  const { node: oldNode, path: commonOldPath } = findTreeDeepestSubcontext(tree, oldContext)
+  const { path: commonNewPath } = findTreeDeepestSubcontext(tree, newContext)
+
+  if (oldNode && commonOldPath.length === oldContext.length) {
+    // if node is already available
+    if (commonOldPath.length === commonNewPath.length) {
+      // if moved with in same place (only rank has changed) (to prevent traversing descendants)
+      nodeChange(tree, oldPath, newPath)
+    }
+    else {
+      if (oldNode.leaf) {
+        nodeChange(tree, newPath.slice(0, newPath.length - 1).concat({ value: '' }), newPath)
+      }
+      else {
+        const descendants = findTreeDescendants(tree, oldContext)
+        descendants.forEach(descendant => {
+          const updatedDescendantPath = newPath.concat(descendant.path.slice(oldPath.length))
+          nodeChange(tree, newPath.slice(0, newPath.length - 1).concat({ value: '' }), updatedDescendantPath)
+        })
+      }
+      nodeDelete(tree, oldPath, false)
+    }
+  }
+  else {
+    // if exact node is not found in the tree
+    nodeChange(tree, newPath.slice(0, newPath.length - 1).concat({ value: '' }), newPath)
+  }
+}
+
+export const treeChange = (originalTree, oldPath, newPath) => produce(originalTree, draftTree => nodeChange(draftTree, oldPath, newPath))
+
+export const treeDelete = (originalTree, oldPath) => produce(originalTree, draftTree => nodeDelete(draftTree, oldPath))
+
+export const treeMove = (originalTree, oldPath, newPath) => produce(originalTree, draftTree => nodeMove(draftTree, oldPath, newPath))
```

Comment:
Using some FP magic, we can abstract the pattern of using immer on the first argument:

```
/** Use immer to pass a draft object as the first argument to the given destructive function to avoid mutation **/
const immerfy = f => (obj, ...rest) => produce(obj, draft => f(draft, ...rest))

export const treeChange = immerfy(nodeChange)
export const treeDelete = immerfy(nodeDelete)
export const treeMove = immerfy(nodeMove)
```

Reply:
Agreed!